docs/docs/09-rfcs/2023-03-08-wing-preview-environment.md

```diff
@@ -0,0 +1,102 @@
+---
+title: Wing Cloud Preview Environments
+description: This document describes the Wing Cloud Preview Environments feature.
+---
+
+# #3696 - Wing Cloud Preview Environments
+
+- **Author(s)**: @ainvoner, @skyrpex
+- **Submission Date**: 2023-08-03
+- **Stage**: Proposal
+- **Stage Date**: 2023-08-03
+
+This document describes the Wing Cloud Preview Environments feature.
+
+## Introduction
+The **Wing Cloud Preview Environments** project is designed to achieve two key objectives. Firstly, it aims to offer a preview environment solution for those interested in using winglang.
+Secondly, it seeks to enhance the winglang ecosystem and its toolchain by building a production application using the language.
+
+## Requirements
+Create the first production feature of Wing Cloud - Wing cloud Preview Environments.
+Build and Deploy wing applications to a Wing Console simulated environment.
+Allow developers to iterate with their teammates using a shared Wing Console instance.
+Seamlessly integrating with GitHub.
+
+### Developer Experience
+A developer using winglang has the capability to integrate preview environments with his/her repository.
+
+#### Sign Up and Installation
+In order to use Wing Cloud Preview Environments, you should sign up to Wing Cloud.
+The signup and installation process is straightforward using GitHub application:
+
+1. Goto https://wing.cloud website and click "Deploy with Wing Cloud"
+2. Complete GitHub authentication.
+3. Grant repository access permission.
+4. You are now signed up and ready to use Wing Cloud Preview Environments.
+
+#### Preview Environment for production branch
+For the production branch, there is a single preview environment that will constantly be up-to-date with the latest code committed.
+A link to the preview environment is available in the repository main page ("About" section).
+
+#### Preview Environments for Pull Requests
+Upon each creation of a pull request, an automatic comment will be added to the PR and will guide the developer to a dedicated preview environment. The deployed preview environment will have a unique url for easy identification.
+To provide a streamlined process for updating a pull request preview environment, each PR code changes will redeploy the preview environment and also will:
+1. provide a real-time build and deployment updates
+2. each entry point will have a seperated preview environment
+3. ensure a consistent url for each preview environment
+4. provide the ability to download the preview environment deployment logs for debugging purposes in case of failure
+
```

Comment:
what happens if the branch is private? how do we ensure other people can't access it? do we use github auth there too?

Reply:
perhaps for our MVP we can assume that all users allowed to view the PR can also view the preview env. and logs.
what do you think?

Reply:
yes, but just bear in mind that this means that access to the environment must be authenticated and there should be some authorization performed against github.

Reply:
accessing env? or creating new ones? even vercel doesn't require authentication for private repositories preview environments.
We had to pay for this service when our console was a private repo.
I don't think we should limit viewing environments.

Reply:
The way Vercel works is what makes most sense IMO:

- GitHub authentication is only used to authorize the Wing Cloud GitHub App to have access to the repository
- The Preview Environment can be private but in order to see it, a user needs to be logged in to Wing Cloud and have view permission to the environment (the permissions are controlled by Wing Cloud, not GitHub anymore)

This is why is makes sense to me:

- A GitHub user having read permission to the repository code doesn't mean they should be able to see the preview environment
- Our auth code won't need to closely work along with GitHub's API, so it'll be much easier to maintain